runtime/src/chpl-visual-debug.c

```diff
@@ -0,0 +1,357 @@
+        }
+      } else {
+        fprintf (stderr, "Can not make Visual Debug directory %s.\n", rootname);
+        return -1;
+      }
+    } else {
+      if ((sb.st_mode & S_IFMT) != S_IFDIR) {
+        fprintf (stderr, "%s: not a directory.\n", rootname);
+        return -1;
+      }
+    }
+    
+    snprintf (fname, 2048, "%s/%s-%d", rootname, rootname, chpl_nodeID);
+    chpl_vdebug_fd = open (fname, O_WRONLY|O_CREAT|O_TRUNC|O_APPEND, 0666);
+    if (chpl_vdebug_fd < 0) {
+      fprintf (stderr, "Visual Debug failed to open %s: %s\n",
+               fname, strerror (errno));
+      chpl_vdebug_fd = -1;
+      return -1;
+    }
+    chpl_vdebug = 1;
+    return 0;
+}
+
+void chpl_vdebug_nolog () {
+  chpl_vdebug = 0;
+}
+
+void chpl_vdebug_start (const char *fileroot, double now) {
+  const char * rootname;
+  struct rusage ru;
+  struct timeval tv;
+  struct timezone tz = {0,0};
+  (void)gettimeofday(&tv, &tz);
+
+  chpl_vdebug = 0;
+
+  // Close any open files.
+  if (chpl_vdebug_fd >= 0)
+    chpl_vdebug_stop ();
+    
+  // Initial call, open file and write initialization information
+  
+  // Get the root of the file name.
+  rootname = (fileroot == NULL || fileroot[0] == 0) ? ".Vdebug" : fileroot; 
+  
+  // In case of an error, just return
+  if (chpl_make_vdebug_file (rootname) < 0)
+    return;
+  
+  // Write initial information to the file, including resource time
+  if ( getrusage (RUSAGE_SELF, &ru) < 0) {
+    ru.ru_utime.tv_sec = 0;
+    ru.ru_utime.tv_usec = 0;
+    ru.ru_stime.tv_sec = 0;
+    ru.ru_stime.tv_usec = 0;
+  }
+  chpl_dprintf (chpl_vdebug_fd,
+                "ChplVdebug: nodes %d id %d seq %.3lf %lld.%06ld %ld.%06ld %ld.%06ld \n",
+                chpl_numNodes, chpl_nodeID, now,
+                (long long) tv.tv_sec, (long) tv.tv_usec,
+                (long)ru.ru_utime.tv_sec, (long)ru.ru_utime.tv_usec,
+                (long)ru.ru_stime.tv_sec, (long)ru.ru_stime.tv_usec  );
+  chpl_vdebug = 1;
+}
+
+void chpl_vdebug_stop (void) {
+  struct rusage ru;  
+  struct timeval tv;
+  struct timezone tz = {0,0};
+  if (chpl_vdebug_fd >= 0) {
+    (void)gettimeofday(&tv, &tz);
+    if ( getrusage (RUSAGE_SELF, &ru) < 0) {
+      ru.ru_utime.tv_sec = 0;
+      ru.ru_utime.tv_usec = 0;
+      ru.ru_stime.tv_sec = 0;
+      ru.ru_stime.tv_usec = 0;
+    }
+    // Generate the End record
+    chpl_dprintf (chpl_vdebug_fd, "End: %lld.%06ld %ld.%06ld %ld.%06ld %d\n",
+                  (long long) tv.tv_sec, (long) tv.tv_usec,
+                  (long)ru.ru_utime.tv_sec, (long)ru.ru_utime.tv_usec,
+                  (long)ru.ru_stime.tv_sec, (long)ru.ru_stime.tv_usec,
+                  chpl_nodeID);
+    close (chpl_vdebug_fd);
+  }
+  chpl_vdebug = 0;
+}
+
+static int tag_no = 0;  // A unique tag number for sorting tags ...
+void chpl_vdebug_tag (const char *str)
+{
+  struct rusage ru;
+  struct timeval tv;
+  struct timezone tz = {0,0};
+
+  (void)gettimeofday(&tv, &tz);
+  if ( getrusage (RUSAGE_SELF, &ru) < 0) {
```

Comment:
Yeah, that chpl_comm_test_nb_complete() arrived with the remote caching and is an example of divergence due to the addition of more developers. :-)     As long as your code is internally consistent I think probably I need to just relax and let the logic flow.   ommmmmm ...